CRM/Contribute/BAO/Contribution.php

```diff
@@ -3539,6 +3539,10 @@ public static function updateFinancialAccounts(&$params, $context = NULL) {
           // & this can be removed
```

Comment:
In general. the updateFinancialAccounts function is difficult to understand as:

1. there are not enough 'business-level' comments in each of these conditions, ie. above line 3481:
// When a pending or partially paid contribution is completed with another payment instrument
// then there is nothing to do as the older contribution should keep the financial accounts it already has.

1. and the conditions overlap by their similarity, ie.:
- it is not immediately clear what is different between lines 3488 and 3510 and 3530,
- or why the structure if the test if not more progressive, ie. outer test on $context only, then test on previousContributionStatus only, and finally test on currentContributionStatus. Concrete example: lines 3528 and 3530 could be deleted since we close an outer if ($context == 'changedStatus') test to re-open the same. Then the structure of tests inside would be clearer.

Reply:
I agree this is a difficult function to understand. There have been a number of improvements in this area over the course of the year - each one with associated unit tests, but there is still a way to go. I do think more people contributing refactorings here would be helpful. 

SImply doing a pull request with comments to go into the code is also a good step